engines/ic3base.cpp

```diff
@@ -346,32 +432,69 @@ bool IC3Base::rel_ind_check(size_t i, const IC3Formula & c, IC3Formula & out)
   solver_->assert_formula(solver_->make_term(Not, c.term));
   // Trans
   assert_trans_label();
-  // c'
-  solver_->assert_formula(ts_.next(c.term));
 
-  Result r = check_sat();
+  // use assumptions for c' so we can get cheap initial
+  // generalization if the check is unsat
+
+  // NOTE: relying on same order between assumps_ and c.children
+  assumps_.clear();
+  {
+    // TODO shuffle assumps and (a copy of) c.children
+    //      if random seed is set
+    Term lbl, ccnext;
+    for (const auto & cc : c.children) {
+      ccnext = ts_.next(cc);
+      lbl = label(ccnext);
+      solver_->assert_formula(solver_->make_term(Implies, lbl, ccnext));
+      assumps_.push_back(lbl);
+    }
+  }
+
+  Result r = check_sat_assuming(assumps_);
   if (r.is_sat()) {
-    if (options_.ic3_pregen_) {
-      out = generalize_predecessor(i, c);
-    } else {
-      out = get_model_ic3formula();
+    if (get_pred) {
+      if (options_.ic3_pregen_) {
+        out = generalize_predecessor(i, c);
```

Comment:
why not pass `out` as a parameter to `generalize_predecessor`?

Reply:
Yeah I think that makes sense. So `get_model_ic3formula` is always called outside of it? The only reason that might not work well is that depending on the generalization approach (e.g. in a derived class) it might need more than just the cube. It might also ask for input / next state values, in which case it would call `get_model_ic3formula` again but with `out_inputs` and `out_nexts` non-null. But, I could refactor it more so that getting the input and next state variable values is a totally separate method. What do you think of that?

Reply:
i think, it is fine the way it is now.

Reply:
Made those changes in 5c176ebd6046ac14f9bc0148c0282f5064ed3ec3. Let me know what you think.

Reply:
lgtm